custom_components/adaptive_lighting/switch.py

```diff
@@ -1689,11 +1739,169 @@ def __init__(self, hass: HomeAssistant):
+                )
+
+                _LOGGER.debug("Proactive adaptation enabled")
+            except RuntimeError:
+                _LOGGER.warning(
+                    "Failed to set up service call interceptors, "
+                    "falling back to event-reactive mode",
+                    exc_info=True,
+                )
+
+    def disable(self):
+        """Disable the listener by removing all subscribed handlers."""
+        for remove in self.listener_removers:
+            remove()
+
+    def set_proactively_adapting(self, context_id: str, entity_id: str) -> None:
+        """Declare the adaptation with the given context ID as proactively adapting,
+        and associate it to an entity ID."""
+        self._proactively_adapting_contexts[context_id] = entity_id
+
+    def is_proactively_adapting(self, context_id: str) -> bool:
+        """Determine whether an adaptation with the given context ID is proactive."""
+        is_proactively_adapting_context = (
+            context_id in self._proactively_adapting_contexts
+        )
+
+        _LOGGER.debug(
+            "is_proactively_adapting_context %s %s",
+            context_id,
+            is_proactively_adapting_context,
+        )
+
+        return is_proactively_adapting_context
+
+    def clear_proactively_adapting(self, entity_id: str) -> None:
+        """Clear all context IDs associated with the given entity ID.
+
+        Call this method to clear past context IDs and avoid a memory leak."""
+        keys = [
+            k for k, v in self._proactively_adapting_contexts.items() if v == entity_id
+        ]
+
+        for key in keys:
+            self._proactively_adapting_contexts.pop(key)
+
+    async def _service_interceptor_turn_on_handler(
+        self, call: ServiceCall, data: ServiceData
+    ):
+        # Don't adapt our own service calls
+        if is_our_context(call.context):
+            return
+
+        entity_ids = self._get_entity_list(data)
+
+        # For simplicity, only service calls affecting a single entity are currently handled.
+        #
+        # To add support for adapting multiple entities, the following properties
+        # need to hold for _all_ entities:
+        # - managed by this AL instance
+        # - not manually controlled
+        # - supporting the same relevant feature set
+        # - off state
+        if len(entity_ids) != 1:
+            return
+
+        entity_id = entity_ids[0]
+
+        if entity_id not in self.adaptive_switch.lights:
+            return
+
+        if self.manual_control.get(entity_id, False):
+            return
+
+        # Prevent adaptation of TURN_ON calls when light is already on,
+        # and of TOGGLE calls when toggling off.
+        if self.hass.states.is_state(entity_id, STATE_ON):
+            return
+
+        _LOGGER.debug(
+            "Intercepted TURN_ON call with data %s (%s)", data, call.context.id
+        )
+
+        self.reset(entity_id, reset_manual_control=False)
```

Comment:
I have been stuck on this for a while.

Do you remember why reset is needed here?

Reply:
I finally got it now :smile: